src/server/scene/basic_surface.cpp

```diff
@@ -165,6 +165,12 @@ void ms::SurfaceObservers::application_id_set_to(Surface const* surf, std::strin
                  { observer->application_id_set_to(surf, application_id); });
 }
 
+ms::BasicSurface::ProofOfMutexLock::ProofOfMutexLock(std::unique_lock<std::mutex> const& lock)
+{
+    if (!lock.owns_lock())
+        BOOST_THROW_EXCEPTION(std::logic_error("ProofOfMutexLock created with unlocked unique_lock"));
```

Comment:
I think this ought to be a `fatal_error`, because what we are checking is the internal logic of `BasicSurface`.

Reply:
I'm open to alternative ways of reporting this error, but not a fan of crashing production builds of Mir unless all hope is lost. If we hit this in production we may need to destroy a surface or disconnect a client, but IMO that shouldn't bring down the whole server along with unrelated windows.

Reply:
We're not checking for unexpected external behaviour. We're double checking our own internal logic (which ought to right). If it is wrong, then we need to fix it long before release.

An occasional lost window or client could be overlooked in testing, especially if intermittent. A nice core will not.

Reply:
Ok. My feeling is that perhaps we should have a mechanism that errors catastrophically in dev/testing and less so in production (though that might be controversial and is *way* out-of-scope for this PR)

Reply:
> Ok. My feeling is that perhaps we should have a mechanism that errors catastrophically in dev/testing and less so in production (though that might be controversial and is _way_ out-of-scope for this PR)

A bit like `fatal_error`? That defaults to `fatal_error_abort` but can be changed to `fatal_error_except` when deployed to systems that want handle exceptions instead?

Reply:
If it doesn't always bring down mir I'm not sure if `fatal_error` is correctly named. I'll ask you how that works next sync.